### PR TITLE
Memoryview serialisation

### PR DESCRIPTION
**Pull request by martindurant**
Fixes https://github.com/dask/distributed/issues/3640 (for discussion)

Does not create an intermediate bytes object for serialization, passes through the memoryview. That, of course, still gets sent as bytes on the wire, so for deserialization, we just wrap it.

### CODE REVIEW

**Comment by mrocklin** [comment]
+1

File: distributed/protocol/serialize.py [hunk 1]
```diff
@@ -568,6 +568,11 @@ def _deserialize_bytes(header, frames):
     return b"".join(frames)
 
 
+@dask_deserialize.register(memoryview)
+def _serialize_memoryview(header, frames):
+    return memoryview(b"".join(frames))
```
Comment (TomAugspurger) [inline distributed/protocol/serialize.py:573]: Is it worth special casing length-1 `frames`, since then you can just return `frames[0]` and avoid the copy? Maybe?

Comment (martindurant) [reply distributed/protocol/serialize.py:573]: Good point

Comment (martindurant) [reply distributed/protocol/serialize.py:573]: Also true for `_deserialize_bytes`, right?

**Comment by jakirkham** [comment]
Thanks Martin!

So one thing I'm curious about is there's a [fair bit of logic in NumPy serialization]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py ), which we seem to be skipping here. For example it is possible to have Python objects much like one can in NumPy arrays. So it might make sense to [pickle in that case]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py#L24-L27 ). Similarly we may want logic to [flatten N-D data]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py#L63-L74 ), [cast larger types]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py#L76-L77 ), and [split large frames]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py#L90-L93 ). Though we [don't need to worry about `datetime` or `timedelta` objects]( https://github.com/dask/distributed/blob/3de9973cafaf85809f712a719111b8db3839975d/distributed/protocol/numpy.py#L79-L83 ) as they are not supported types in `memoryview`s. Given this, what do you think about repurposing this logic for `memoryview`s and allowing NumPy serialization to just piggyback on `memoryview` serialization?

**Comment by martindurant** [comment]
> t is possible to have Python objects

I am surprised to find that `memoryview.format == "O"` is indeed allowed. Does this actually ever happen? I suppose this just failed previously, or it never ever happened, but I am not sure.

I think the same goes for complex types and multiple dimensions - those should be already using numpy. I daresay that in every case here, we have a bytes-like thing (from arrow, whatever) that we don't want to accidentally copy.

In short, I am not keen to change the numpy side of things, even if it just amounts to moving code around a little, to support memoryviews that I'd be surprised if they got used. Prepard to be proven wrong...

**Comment by jakirkham** [comment]
I think this goes back to the fact that the Python Buffer Protocol allows things like arrays of pointers. This was to make PIL developers happy with the PEP IIRC. So `memoryview.format == "O"` is one such case as one would have `PyObject*` in the array. In practice, this might come up when serializing a column of strings? 🤷‍♂️

Was mostly thinking we could get more out of the same code. Though you may be right this seldom comes up in practice.

Yeah I guess we wait and see if there's an issue. 🙂

**Comment by martindurant** [comment]
>  In practice, this might come up when serializing a column of strings?

Since pandas and numpy already have their own serialisation, someone would *really* have to be trying to break Dask :)
